Suggest changes to two quirky-looking bits

dougbu opened the pull request:
Update two quirky-looking bits of code
- use SOAP fault when Salesforce handler fails
  - preserve non-200 but successful HTTP status code
- add lots of comments about `IsAssignableFrom(...)` use in `GetDataOrDefault<T>(...)`
  - test this method more thoroughly

nits:
- use new C# features
- remove duplicate test data in `GenericJsonWebHookReceiverTests`
- take VS suggestions, especially in `WebHookHandlerContextExtensionsTests`

dnfclas commented:
@dougbu,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

dougbu commented:
Rebased on `dougbu/infrastructure.updates` and changed questions into proposed answers. Bit easier to review with `files?w=1` in the URL.

@HenrikFrystykNielsen I would appreciate your thoughts on the `SalesforceSoapWebHookReceiver` and `WebHookHandlerContextExtensions` changes. Am I missing something? And, why aren't the unit tests affected? (Should I add more tests?)

dougbu commented:
Reopening to see if AppVeyor config is working

dnfclas commented:
@dougbu,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

dougbu commented:
@HenrikFrystykNielsen there's now nothing in this PR except my suggestions (and a few trailing whitespace removals). Please have a look&hellip; I am not confident these changes are correct for the existing receivers.

dougbu commented on src/Microsoft.AspNet.WebHooks.Receivers.Salesforce/WebHooks/SalesforceSoapWebHookReceiver.cs:
```diff
@@ -93,12 +93,17 @@ public override string Name
+                    {
+                        var success = ReadResource("Microsoft.AspNet.WebHooks.Messages.NotificationResponse.xml");
+                        response = GetXmlResponse(request, statusCode, success);
+                    }
```
Add `else` to handle the failure case w/ the SOAP error response.

dougbu commented on src/Microsoft.AspNet.WebHooks.Receivers/Extensions/WebHookHandlerContextExtensions.cs:
```diff
@@ -30,8 +30,10 @@ public static T GetDataOrDefault<T>(this WebHookHandlerContext context)
             }
 
-            if (context.Data is JToken && !typeof(JToken).IsAssignableFrom(typeof(T)))
+            if (context.Data is JToken && !typeof(T).IsAssignableFrom(context.Data.GetType()))
```
Add tests to confirm potential failure cases e.g. `T` is `JArray` but the current object is `JObject`.

dougbu replied:
Testing showed this change only affects corner cases. It's less efficient in some of those cases e.g. where `Data` is a `JObject` and `T` is `JArray`. Will attempt the impossible conversion and catch the thrown `Exception` rather than simply returning `null`. Don't see a compelling reason to do more than add comments and tests.

dougbu commented:
With thanks for his offline comments, removing @HenrikFrystykNielsen's review request.

dougbu commented:
:up::date: and ready for @rynowak

rynowak commented on src/Microsoft.AspNet.WebHooks.Receivers.Salesforce/WebHooks/SalesforceSoapWebHookReceiver.cs:
```diff
@@ -93,12 +93,28 @@ public override string Name
+                        // that sensible choice into this SOAP 1.1 implementation).
+                        var resource = ReadResource("Microsoft.AspNet.WebHooks.Messages.FaultResponse.xml");
+                        var faultString = $"Handler's status: {statusCode}, '{response.ReasonPhrase}'.";
+                        var failure = string.Format(CultureInfo.InvariantCulture, resource, faultString);
```
Should this be in a resource since it's going out of the wire?

dougbu replied:
Which "this"? If you mean the `faultString`, perhaps. I'll also check other places we use the `FaultResponse.xml` resource.

rynowak replied:
Yeah I just wondered. Even if we don't do loc it's nice to have a separation between strings that a user sees and strings that are just used as IDs.

rynowak commented on src/Microsoft.AspNet.WebHooks.Receivers/Extensions/WebHookHandlerContextExtensions.cs:
```diff
@@ -30,16 +30,34 @@ public static T GetDataOrDefault<T>(this WebHookHandlerContext context)
+            if (context.Data is JToken token && !typeof(JToken).IsAssignableFrom(typeof(T)))
             {
+                // context.Data is a JToken and the caller isn't just asking for (say) the current JObject.
+                // Do explicit conversion.
```
So there's this huge comment here explaining the design tradeoffs but what does this code **mean**? Does it basically say that this block handles the case where T is one of the JLinq types?

dougbu replied:
The block handles the case where `Data` is one of the JLinq types and `T` is *not* e.g. `Data` is a `JObject` and `T` is something like `AzureAlertNotification` (one of the strong types for request content). I'll add this info before the design "discussion".

dougbu commented:
78491ef819
